packages/webchannel-wrapper/src/index.d.ts

```diff
@@ -15,23 +15,65 @@
  * limitations under the License.
  */
 
-// TODO(mikelehen): Flesh out proper types (or figure out how to generate via
-// clutz or something).
-export class XhrIo {}
-export const ErrorCode: any;
-export const EventType: any;
+// WARNING: This is not a complete set of types exported by WebChannelWrapper.
+// It is merely meant to support the usage patterns of the Firestore SDK.
+
+export var EventType: {
+  COMPLETE: string;
+};
+
 export namespace WebChannel {
-  export type EventType = any;
-  export const EventType: any;
+  export var EventType: {
+    OPEN: string;
+    CLOSE: string;
+    ERROR: string;
+    MESSAGE: string;
+  };
+}
+
+export var ErrorCode: {
+  NO_ERROR: number;
+  HTTP_ERROR: number;
+  TIMEOUT: number;
+};
+
+export interface Headers {
+  [name: string]: string | number;
+}
+
+export interface WebChannelError {
+  error?: { status: string; message: string };
 }
 
-type StringMap = { [key: string]: string };
+export class XhrIo {
+  send(
+    url: string,
+    method?: string,
+    body?: string,
+    headers?: Headers,
+    timeoutInterval?: number
+  ): void;
+
+  getLastErrorCode(): number;
+
+  getLastError(): string;
+
+  getStatus(): number;
+
+  getResponseText(): string;
+
+  getResponseJson(): WebChannelError | object;
+
+  listenOnce<T>(type: string, cb: (param: T) => void): void;
```

Comment:
same feedback about generics here.

Reply:
Removed. I am "casting" in the callsite now.